### PR TITLE
feat: toolbar updates for token approvals

### PR DESCRIPTION
**Pull request by just-toby**
- change placement of tooltips relative to the anchoring element
- change the color of the Toolbar content for "failed approval" state - yellow/warning color
- some layout updates to Toolbar to accommodate small widths
- change the spinner icon to match design

### CODE REVIEW

File: src/assets/svg/spinner.svg [hunk 1]
```diff
@@ -1,17 +1,7 @@
...
-        cy="12"
-        r="10"
-        stroke-width="2"
-        stroke-linecap="round"
-        stroke-linejoin="round"
-        mask="url(#mask)"
-      />
-    </svg>
\ No newline at end of file
+<svg viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (zzmp) [inline src/assets/svg/spinner.svg:1]: Why this change? The spinner won't render on GitHub, but the current state is explicitly a circle, while the new state is a path, so it may render worse if magnified.

File: src/components/Swap/Toolbar/index.tsx [hunk 2]
```diff
@@ -16,8 +16,10 @@ import * as Caption from './Caption'
 const ToolbarRow = styled(Row)`
   border: 1px solid ${({ theme }) => theme.outline};
   border-radius: ${({ theme }) => theme.borderRadius - 0.25}em;
+  flex-wrap: nowrap;
   gap: 0.5em;
   min-height: 3.5em;
+  overflow: hidden;
```
Comment (zzmp) [inline src/components/Swap/Toolbar/index.tsx:22]: In which case is it overflowing?

File: src/icons/index.tsx [hunk 3]
```diff
@@ -18,12 +16,12 @@ import {
   ArrowDown as ArrowDownIcon,
   ArrowRight as ArrowRightIcon,
   ArrowUp as ArrowUpIcon,
+  ArrowUpRight as LinkIcon,
```
Comment (zzmp) [inline src/icons/index.tsx:19]: Unused - remove?

File: src/components/Swap/Toolbar/Caption.tsx [hunk 4]
```diff
@@ -136,7 +150,7 @@ export interface TradeProps extends GasEstimateProps {
 export function Trade({ trade, outputUSDC, impact, gasUseEstimateUSD }: TradeProps) {
   return (
     <>
-      <CaptionRow gap={0.5}>
+      <CaptionRow gap={0.5} shrink={0}>
```
Comment (zzmp) [inline src/components/Swap/Toolbar/Caption.tsx:153]: Could `flex-shrink: 0` be defined in `CaptionRow` instead of expanding `Row` to include it for a one-off?

File: src/components/Swap/Toolbar/Caption.tsx [hunk 5]
```diff
@@ -29,36 +30,48 @@ const CaptionRow = styled(Row)<{ gap: number }>`
...
+      {isWideWidget ? (
+        <Gas color={'secondary'} />
+      ) : (
+        <Tooltip icon={Gas} placement="left" iconProps={{ color: 'secondary' }}>
+          <ThemedText.Body2 color="secondary">
+            Estimated gas: {formatCurrencyAmount({ amount: gasUseEstimateUSD, isUsdPrice: true })}
+          </ThemedText.Body2>
+        </Tooltip>
+      )}
+      {isWideWidget && (
```
Comment (zzmp) [inline src/components/Swap/Toolbar/Caption.tsx:62]: Move this into the ternary:

```
      {isWideWidget ? (
        <>
          <Gas color="secondary" />
          <ThemedText.Body2 color='secondary>
            ...
        </>
      ) : (
```

File: src/components/Swap/Toolbar/Caption.tsx [hunk 6]
```diff
@@ -29,36 +30,48 @@ const CaptionRow = styled(Row)<{ gap: number }>`
...
 function GasEstimate({ gasUseEstimateUSD }: GasEstimateProps) {
+  const isWideWidget = useIsWideWidget()
   return (
     <CaptionRow gap={0.25}>
-      <Gas color="secondary" />
-      <ThemedText.Body2 color="secondary">
-        {formatCurrencyAmount({ amount: gasUseEstimateUSD, isUsdPrice: true })}
-      </ThemedText.Body2>
+      {isWideWidget ? (
+        <Gas color={'secondary'} />
```
Comment (zzmp) [inline src/components/Swap/Toolbar/Caption.tsx:54]: `color="secondary"`

File: src/components/Swap/SwapActionButton/AllowanceButton.tsx [hunk 7]
```diff
@@ -47,28 +50,39 @@ export default function AllowanceButton({ token, isApprovalLoading, approveAndPe
...
+          ? t`Approve ${token?.symbol ?? 'token'} for trading`
+          : t`Approve ${token?.symbol ?? 'token'}`,
         onClick,
       }
     }
-  }, [isApprovalLoading, isFailed, isPending, onClick, pendingApproval, token?.symbol])
+  }, [isApprovalLoading, isFailed, isPending, isWideWidget, onClick, pendingApproval, token?.symbol])
+
+  const defaultButtonColor = useTokenColorExtraction()
+  const buttonColor = useMemo(() => {
```
Comment (zzmp) [inline src/components/Swap/SwapActionButton/AllowanceButton.tsx:77]: nit: one-line as a ternary:

```
const buttonColor = useMemo(() => isFailed ? 'warningSoft' : defaultButtonColor, [defaultButtonColor, isFailed])
```

File: src/components/Swap/SwapActionButton/AllowanceButton.tsx [hunk 8]
```diff
@@ -47,28 +50,39 @@ export default function AllowanceButton({ token, isApprovalLoading, approveAndPe
...
-        message: t`Approval failed`,
+        message: t`${token?.symbol ?? 'Token'} approval failed`,
         onClick,
+        color: 'warning',
       }
     } else {
       return {
-        tooltipContent: t`Permission is required for Uniswap to swap each token. This will expire after one month for your security.`,
-        message: t`Approve use of ${token?.symbol ?? 'token'}`,
+        tooltipContent: t`Permission is required for Uniswap smart contract to use each token. This will expire after one month for your security.`,
```
Comment (zzmp) [inline src/components/Swap/SwapActionButton/AllowanceButton.tsx:67]: Copy needs work: it should either be "for Uniswap to" or "for **the** Uniswap smart contract to". As-is is grammatically off.

File: src/components/Swap/SwapActionButton/AllowanceButton.tsx [hunk 9]
```diff
@@ -36,7 +37,9 @@ export default function AllowanceButton({ token, isApprovalLoading, approveAndPe
     }
   }, [approveAndPermit])
 
-  const action = useMemo(() => {
+  const isWideWidget = useIsWideWidget()
+
```
Comment (zzmp) [inline src/components/Swap/SwapActionButton/AllowanceButton.tsx:41]: nit: no newline. it's part of the same business logic.

File: src/components/ActionButton.tsx [hunk 10]
```diff
@@ -68,11 +67,14 @@ export interface Action {
   icon?: Icon
   tooltipContent?: ReactNode
   onClick?: () => void
+  color?: Color
   children?: ReactNode
 }
 
+export type ActionButtonColor = keyof Pick<Colors, 'accent' | 'accentSoft' | 'warningSoft' | 'interactive'>
```
Comment (zzmp) [inline src/components/ActionButton.tsx:74]: This does not need to be exported.

File: src/components/ActionButton.tsx [hunk 11]
```diff
@@ -68,11 +67,14 @@ export interface Action {
...
+
 export interface BaseProps {
```
Comment (zzmp) [inline src/components/ActionButton.tsx:76]: I don't think this needs to be exported.

**Review by zzmp — CHANGES_REQUESTED** [review]

File: src/components/Swap/SwapActionButton/AllowanceButton.tsx (hunk 8 above)
Comment (just-toby) [reply src/components/Swap/SwapActionButton/AllowanceButton.tsx]: i'll add `the` - copied this directly from the figma

File: src/components/Swap/Toolbar/Caption.tsx (hunk 4 above)
Comment (just-toby) [reply src/components/Swap/Toolbar/Caption.tsx:154]: sure - any reason we don't want to expand `Row`?

File: src/components/Swap/Toolbar/index.tsx (hunk 2 above)
Comment (just-toby) [reply src/components/Swap/Toolbar/index.tsx:22]: when the width is really small and I add "flex-wrap", the content started overflowing. i don't think we want multi-line wrapping so i'm taking the approach of truncating the content to fit in the smaller width

File: src/icons/index.tsx (hunk 3 above)
Comment (just-toby) [reply src/icons/index.tsx:19]: this is still used, i'm just switching the underlying icon

**Review by zzmp — CHANGES_REQUESTED** [review]

**Review by zzmp — CHANGES_REQUESTED** [review]
Why did you change the spinner.svg? Considering that it's no longer a circle (but a path), could it be changed back?

File: src/assets/svg/spinner.svg (hunk 1 above)
Comment (zzmp) [reply src/assets/svg/spinner.svg:1]: it is a new image, so it's worth the change. it _could_ be implemented as a more human-readable svg, but is not worth the effort in this case.

**Review by zzmp — APPROVED** [review]
